subsys/mgmt/mcumgr/grp/fs_mgmt/src/fs_mgmt.c

```diff
@@ -354,44 +398,186 @@ static int fs_mgmt_file_upload(struct smp_streamer *ctxt)
 			return MGMT_ERR_EINVAL;
 		}
+	}
 
-		fs_mgmt_ctxt.uploading = true;
-		fs_mgmt_ctxt.off = 0;
-		fs_mgmt_ctxt.len = len;
-	} else {
-		if (!fs_mgmt_ctxt.uploading) {
-			return MGMT_ERR_EINVAL;
+	if (k_sem_take(&fs_mgmt_ctxt.lock_sem, FILE_SEMAPHORE_MAX_TAKE_TIME)) {
+		return MGMT_ERR_EBUSY;
+	}
+
+	/* Check if there is an active upload in progress */
+	if (ctxt->smpt != fs_mgmt_ctxt.transport) {
+		/* Different transport, clean up */
+		close_file = true;
+		open_file = true;
+	} else if (fs_mgmt_ctxt.state == STATE_UPLOAD) {
+		/* Check if this is the same file */
+		if (strcmp(file_name, fs_mgmt_ctxt.path)) {
+			/* Different file, we need to clean up */
+			close_file = true;
+			open_file = true;
 		}
+	} else if (fs_mgmt_ctxt.state == STATE_DOWNLOAD) {
+		/* Different mode, we need to clean up */
+		close_file = true;
+		open_file = true;
+	} else {
+		open_file = true;
+	}
+
+	/* Close existing file if it was open */
+	if (close_file) {
+		fs_mgmt_cleanup();
+	}
+
+	/* Open new file */
+	if (open_file) {
+		fs_mgmt_ctxt.off = 0;
+		fs_file_t_init(&fs_mgmt_ctxt.file);
+		rc = fs_open(&fs_mgmt_ctxt.file, file_name, FS_O_CREATE | FS_O_WRITE);
 
-		if (off != fs_mgmt_ctxt.off) {
-			/* Invalid offset.  Drop the data and send the expected offset. */
-			return fs_mgmt_file_rsp(zse, MGMT_ERR_EINVAL, fs_mgmt_ctxt.off);
+		if (rc != 0) {
+			k_sem_give(&fs_mgmt_ctxt.lock_sem);
+			return MGMT_ERR_ENOENT;
 		}
+
+		strcpy(fs_mgmt_ctxt.path, file_name);
+		fs_mgmt_ctxt.state = STATE_UPLOAD;
+		fs_mgmt_ctxt.transport = ctxt->smpt;
+	}
+
+	if (off == 0) {
+		fs_mgmt_ctxt.len = len;
 	}
 
-	new_off = fs_mgmt_ctxt.off + file_data.len;
+	new_off = off + file_data.len;
+
 	if (new_off > fs_mgmt_ctxt.len) {
-		/* Data exceeds image length. */
-		return MGMT_ERR_EINVAL;
+		/* Data exceeds image length, but check if this is because we need to re-create
+		 * the variable.
+		 */
+		if (fs_mgmt_ctxt.len == 0) {
+			rc = fs_mgmt_filelen(file_name, &fs_mgmt_ctxt.len);
+			if (rc != 0) {
+				fs_mgmt_cleanup();
+				k_sem_give(&fs_mgmt_ctxt.lock_sem);
+				return MGMT_ERR_EUNKNOWN;
+			}
+		} else {
+			fs_mgmt_cleanup();
+			k_sem_give(&fs_mgmt_ctxt.lock_sem);
+			return MGMT_ERR_EINVAL;
+		}
 	}
 
 	if (file_data.len > 0) {
 		/* Write the data chunk to the file. */
-		rc = fs_mgmt_write(file_name, off, file_data.value, file_data.len);
-		if (rc != 0) {
-			return rc;
+		size_t file_size = 0;
+		bool reset_position = false;
+
+		if (off == 0) {
+			rc = fs_mgmt_filelen(file_name, &file_size);
+
+			if (fs_mgmt_ctxt.off > 0) {
+				/* Position in file needs to be reset due to truncation */
+				reset_position = true;
+			}
+
+			fs_mgmt_ctxt.off = 0;
+		} else if (off != fs_mgmt_ctxt.off) {
+			/* Seek to desired offset */
+			rc = fs_seek(&fs_mgmt_ctxt.file, off, FS_SEEK_SET);
+
+			if (rc != 0) {
+				fs_mgmt_cleanup();
+				k_sem_give(&fs_mgmt_ctxt.lock_sem);
+				return MGMT_ERR_EUNKNOWN;
+			}
+
+			fs_mgmt_ctxt.off = off;
+		}
```

Comment:
Do you mean to skip parts of an upload, e.g. if a user uploads a 4KB file and from 2K-3K it is 0xff so those bytes should be skipped? I guess the problem with that is that the actual data that would be there would be undefined, would it be 0xff or 0x00? Only the filesystem knows that. Would also need to check that you can seek to an offset that doesn't exist and that the file would be expanded, not sure if you would need to truncate the file to the new size first or not. On the other hand, what if there is missed packets that should be in the uploaded file, or would that not matter because the client would get a response

Reply:
No, what I wanted to write is to not extend file, append to offset there is or report error that other offset has been expected (current length of file). I see no reason to skip to some offset. When I have stated "upload empty file" what I have meant is that you could upload chunk 0-2 then chunk 4095-4096 and get empty 40096 byte file by just uploading few bytes.